### PR TITLE
[CDAP-16740] Centralizing logic for EL Context and improve tests

### PR DESCRIPTION
**Pull request by nitinmotgi**
It looks like a lot of files, but those are from refactoring of changing Row.length() to Row.width(). Internal function.

### CODE REVIEW

File: wrangler-api/src/main/java/io/cdap/wrangler/api/Row.java [hunk 1]
```diff
@@ -170,9 +170,9 @@ public int find(String col) {
   }
 
   /**
-   * @return  Length of the row.
+   * @return  width of the row.
    */
-  public int length() {
+  public int width() {
```
Comment (CuriousVini) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/Row.java:183]: This is a change in api. Is this change backwards compatible? would this require redeployment of existing pipelines?

File: wrangler-api/src/main/java/io/cdap/wrangler/api/ErrorRowException.java [hunk 2]
```diff
@@ -42,7 +39,7 @@ public ErrorRowException(String message, int code) {
   }
 
   public ErrorRowException(String directiveName, String errorMessage, int code) {
-    this(String.format("Error encountered while executing '%s' : %s", directiveName, errorMessage), code);
+    this(String.format("%s (ecode: %d, directive: %s)", errorMessage, code, directiveName), code);
```
Comment (CuriousVini) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/ErrorRowException.java:42]: how about if we first have directive name, error message and then error code?

File: wrangler-core/src/main/java/io/cdap/wrangler/expression/EL.java [hunk 3]
```diff
@@ -74,7 +74,7 @@ public void compile(String expression) throws ELException {
       throw new ELException(
         String.format("Error encountered while executing '%s' at line '%d' and column '%d'. " +
                         "Make sure a valid jexl transformation is provided.",
-                      info.getDetail().toString(), info.getLine(), info.getColumn()), e);
+                      info.toString(), info.getLine(), info.getColumn()), e);
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/wrangler/expression/EL.java]: As per the javadoc, getDetail() provides actual part of the code that triggered the error. why do we need this change?

File: wrangler-core/src/main/java/io/cdap/wrangler/executor/RecipePipelineExecutor.java [hunk 4]
```diff
@@ -122,28 +122,31 @@ public void destroy() {
...
+          results.addAll(cumulativeRows);
         } catch (ErrorRowException e) {
           messages.add(String.format("%s", e.getMessage()));
           collector
-            .add(new ErrorRecord(newRows.get(0), String.join(",", messages), e.getCode(),
+            .add(new ErrorRecord(rows.subList(i, i + 1).get(0), String.join(",", messages), e.getCode(),
               e.isShownInWrangler()));
         }
-        i++;
+        ++i;
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/wrangler/executor/RecipePipelineExecutor.java:149]: why is this change needed?

File: wrangler-core/src/test/java/io/cdap/wrangler/ExperimentTests.java [hunk 5]
```diff
@@ -65,13 +61,4 @@ public void testIdCreationFromName() throws Exception {
     }
   }
 
-  @Test
-  public void testCurrencyParsing() throws Exception {
```
Comment (CuriousVini) [inline wrangler-core/src/test/java/io/cdap/wrangler/ExperimentTests.java:69]: why is this removed?

File: wrangler-core/src/test/java/io/cdap/directives/row/RowRegexFilterTest.java [hunk 6]
```diff
@@ -91,7 +91,7 @@ public void testFilterOnNumericValues() throws Exception {
     directives.add("set-type :long long");
     directives.add("set-type :float float");
     directives.add("set-type :double double");
-    directives.add("filter-rows-on condition-false int =^ \"5\"");
+    directives.add("filter-rows-on condition-false int == 5005");
```
Comment (CuriousVini) [inline wrangler-core/src/test/java/io/cdap/directives/row/RowRegexFilterTest.java:94]: why is this change needed?

File: wrangler-core/src/main/java/io/cdap/wrangler/expression/ELContext.java [hunk 7]
```diff
@@ -36,6 +37,23 @@
 public class ELContext implements JexlContext {
   private final Map<String, Object> values = new HashMap<>();
 
+  /**
+   * Context object passed to every expresson evaluation.
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/wrangler/expression/ELContext.java]: typo: `expression`

File: wrangler-core/src/main/java/io/cdap/wrangler/expression/ELContext.java [hunk 8]
```diff
@@ -36,6 +37,23 @@
...
+   * All properties of this class are public to ensure they can be accessed with dot(.) format.
+   */
+  public static class Context {
+    public String environment;
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/wrangler/expression/ELContext.java]: why not have these in the ELContext itself?

File: wrangler-core/src/main/java/io/cdap/wrangler/expression/ELContext.java [hunk 9]
```diff
@@ -45,24 +63,40 @@ public ELContext() {
...
+    // because, el variables might not be present in a row.
+    init(context);
+    set("this", row);
+  }
 
-    values.put("runtime", context.getProperties());
-    values.put("environment", context.getEnvironment().name());
-    values.put("context", context.getContextName());
+  private void init(ExecutorContext context) {
+    if (context != null) {
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/wrangler/expression/ELContext.java:96]: when can context be null? If context can be null it should be marked as nullable in caller as well as in this method

File: wrangler-core/src/main/java/io/cdap/wrangler/expression/ELContext.java [hunk 10]
```diff
@@ -36,6 +37,23 @@
 public class ELContext implements JexlContext {
   private final Map<String, Object> values = new HashMap<>();
 
+  /**
+   * Context object passed to every expresson evaluation.
+   * All properties of this class are public to ensure they can be accessed with dot(.) format.
+   */
+  public static class Context {
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/wrangler/expression/ELContext.java:45]: If this class is accessed by other classes outside this class, why not make it top level class? why should it be nested?

File: wrangler-core/src/main/java/io/cdap/directives/aggregates/SetTransientVariable.java [hunk 11]
```diff
@@ -82,18 +82,7 @@ public void destroy() {
   public List<Row> execute(List<Row> rows, ExecutorContext context) throws DirectiveExecutionException {
     for (Row row : rows) {
       // Move the fields from the row into the context.
-      ELContext ctx = new ELContext(context);
```
Comment (CuriousVini) [inline wrangler-core/src/main/java/io/cdap/directives/aggregates/SetTransientVariable.java:85]: If this is moved to EL class, does it mean custom directives will need to remove this if they are using it and recompile?

**Review by CuriousVini — COMMENTED** [review]
A few questions.

File: wrangler-api/src/main/java/io/cdap/wrangler/api/Row.java (hunk 1 above)
Comment (nitinmotgi) [reply wrangler-api/src/main/java/io/cdap/wrangler/api/Row.java:183]: no, this is internal only. Now, we are exposing it.

File: wrangler-api/src/main/java/io/cdap/wrangler/api/ErrorRowException.java (hunk 2 above)
Comment (nitinmotgi) [reply wrangler-api/src/main/java/io/cdap/wrangler/api/ErrorRowException.java:42]: Directive is secondary information, I would like to know what the error is first and then find out where it's coming from. That's why I changed the order.

File: wrangler-core/src/main/java/io/cdap/wrangler/expression/EL.java (hunk 3 above)
Comment (nitinmotgi) [reply wrangler-core/src/main/java/io/cdap/wrangler/expression/EL.java]: Depending on the situation `getDetail` can be null sometimes. So u can hit `NulPointerException`.

File: wrangler-core/src/main/java/io/cdap/wrangler/executor/RecipePipelineExecutor.java (hunk 4 above)
Comment (nitinmotgi) [reply wrangler-core/src/main/java/io/cdap/wrangler/executor/RecipePipelineExecutor.java:149]: it's just a habit from old days.

File: wrangler-core/src/test/java/io/cdap/wrangler/ExperimentTests.java (hunk 5 above)
Comment (nitinmotgi) [reply wrangler-core/src/test/java/io/cdap/wrangler/ExperimentTests.java:69]: Was wrong test.

File: wrangler-core/src/test/java/io/cdap/directives/row/RowRegexFilterTest.java (hunk 6 above)
Comment (nitinmotgi) [reply wrangler-core/src/test/java/io/cdap/directives/row/RowRegexFilterTest.java:94]: Was a bad test so fixed it.

File: wrangler-core/src/main/java/io/cdap/wrangler/expression/ELContext.java (hunk 7 above)
Comment (nitinmotgi) [reply wrangler-core/src/main/java/io/cdap/wrangler/expression/ELContext.java]: done

File: wrangler-core/src/main/java/io/cdap/wrangler/expression/ELContext.java (hunk 9 above)
Comment (nitinmotgi) [reply wrangler-core/src/main/java/io/cdap/wrangler/expression/ELContext.java:96]: during test

Comment (nitinmotgi) [reply wrangler-core/src/main/java/io/cdap/wrangler/expression/ELContext.java:96]: done

File: wrangler-core/src/main/java/io/cdap/directives/aggregates/SetTransientVariable.java (hunk 11 above)
Comment (nitinmotgi) [reply wrangler-core/src/main/java/io/cdap/directives/aggregates/SetTransientVariable.java:85]: These are advanced, I haven't see any one us these type. EL is not exposed outside.

**Comment by nitinmotgi** [comment]
@CuriousVini PTAL

File: wrangler-api/src/main/java/io/cdap/wrangler/api/Row.java (hunk 1 above)
Comment (albertshau) [reply wrangler-api/src/main/java/io/cdap/wrangler/api/Row.java:183]: This is a backwards incompatible change and will break any existing UDD. We cannot remove any existing method without adding its replacement and deprecating it for a full major release.

In this case, this rename seems unnecessary and should not be done. If it's unclear, just add clearer javadocs.

File: wrangler-core/src/main/java/io/cdap/wrangler/expression/EL.java (hunk 3 above)
Comment (albertshau) [reply wrangler-core/src/main/java/io/cdap/wrangler/expression/EL.java]: I looked at JexlInfo.toString(), this will result in a weird error message with duplicate information:

```
Error encountered while executing '@<line>:<column> ![<detail.start>,<detail.end>]: '<detail.toString>'' at line '<line>' and column '<column>'
```

Is this really what you intended?

File: wrangler-core/src/main/java/io/cdap/wrangler/expression/ELContext.java [hunk 12]
```diff
@@ -36,6 +38,23 @@
...
 
+  /**
+   * Context object passed to every expression evaluation.
+   * All properties of this class are public to ensure they can be accessed with dot(.) format.
+   */
+  public static class Context {
+    public String environment;
+    public String name;
+    public long nano;
+    public long millis;
```
Comment (albertshau) [inline wrangler-core/src/main/java/io/cdap/wrangler/expression/ELContext.java]: don't think it makes sense for an expression to modify any of these right? If so, make these final otherwise directives can mess each other up.

File: wrangler-core/src/main/java/io/cdap/wrangler/expression/ELContext.java [hunk 13]
```diff
@@ -36,6 +38,23 @@
...
+    public Context (String environment, String name) {
```
Comment (albertshau) [inline wrangler-core/src/main/java/io/cdap/wrangler/expression/ELContext.java]: style: remove extra space after Context

File: wrangler-core/src/test/java/io/cdap/directives/aggregates/SetTransientVariableTest.java [hunk 14]
```diff
@@ -50,6 +50,7 @@ public void testSettingOfVariable() throws Exception {
     rows.add(new Row("fwd", 1).add("A", 2));
 
     final Map<String, Object> s = new HashMap<>();
+    s.put("test", 1);
```
Comment (albertshau) [inline wrangler-core/src/test/java/io/cdap/directives/aggregates/SetTransientVariableTest.java]: is this intended? don't see any change to the rest of the test

File: wrangler-core/src/test/java/io/cdap/directives/parser/CsvParserTest.java [hunk 15]
```diff
@@ -67,4 +67,61 @@ public void testHeaders() throws Exception {
...
+      new Row("body", "14J456,65765566,Granger,Hermione,,,"),
+      new Row("body", "14J456,13233121,Diggory,Cedric,,,"),
+      new Row("body", "14J456,98786868,Weasley,George,,,"),
+      new Row("body", "14J456,78977876,Weasley,Fred,,,")
+    );
+
+    rows = TestingRig.execute(directives, rows);
+    Assert.assertEquals(6, rows.size());
+    for (int i = 0; i < rows.size(); ++i) {
+      Assert.assertEquals(4, rows.get(i).width());
```
Comment (albertshau) [inline wrangler-core/src/test/java/io/cdap/directives/parser/CsvParserTest.java:109]: tests should basically never just assert on a size. They should always verify the correct values are being generated.

File: wrangler-core/src/test/java/io/cdap/directives/row/SendToErrorAndContinueTest.java [hunk 16]
```diff
@@ -31,6 +31,36 @@
...
+    );
+
+    RecipePipeline pipeline = TestingRig.execute(directives);
+    List<Row> results = pipeline.execute(rows);
+    List<ErrorRecord> errors = pipeline.errors();
+
+    Assert.assertEquals(2, errors.size());
+    Assert.assertEquals(1, results.size());
+    Assert.assertEquals("1", errors.get(0).getRow().getValue("C"));
+    Assert.assertEquals("filtering records that dont have 4 columns (ecode: 1)", errors.get(0).getMessage());
```
Comment (albertshau) [inline wrangler-core/src/test/java/io/cdap/directives/row/SendToErrorAndContinueTest.java]: these types of assertions are very brittle and easily broken. For example, "dont" should actually be "don't" or "do not". Fixing it would break this test.

File: wrangler-core/src/main/java/io/cdap/directives/row/RecordConditionFilter.java [hunk 17]
```diff
@@ -91,19 +91,7 @@ public void destroy() {
     List<Row> results = new ArrayList<>();
     for (Row row : rows) {
       // Move the fields from the row into the context.
-      ELContext ctx = new ELContext(context);
-      for (String var : el.variables()) {
-        Object value = row.getValue(var);
-        String strValue;
-        // support numeric values by converting them to string
-        if (value instanceof Number) {
-          Number number = (Number) value;
```
Comment (albertshau) [inline wrangler-core/src/main/java/io/cdap/directives/row/RecordConditionFilter.java:100]: Looks like this number conversion will no longer happen.  I don't know why numbers had to be converted to strings to begin with, but I imagine this change will break existing pipelines.

Comment (albertshau) [reply wrangler-core/src/main/java/io/cdap/directives/row/RecordConditionFilter.java:100]: is there test coverage for number input to this directive?

File: wrangler-core/src/main/java/io/cdap/wrangler/expression/ELContext.java (hunk 12 above)
Comment (nitinmotgi) [reply wrangler-core/src/main/java/io/cdap/wrangler/expression/ELContext.java]: done!

File: wrangler-core/src/test/java/io/cdap/directives/aggregates/SetTransientVariableTest.java (hunk 14 above)
Comment (nitinmotgi) [reply wrangler-core/src/test/java/io/cdap/directives/aggregates/SetTransientVariableTest.java]: unintended.

File: wrangler-core/src/test/java/io/cdap/directives/parser/CsvParserTest.java (hunk 15 above)
Comment (nitinmotgi) [reply wrangler-core/src/test/java/io/cdap/directives/parser/CsvParserTest.java:109]: This is testing if all columns are correct or not. The intent of test was to check for right columns. anyway added each column test.

File: wrangler-core/src/test/java/io/cdap/directives/row/SendToErrorAndContinueTest.java (hunk 16 above)
Comment (nitinmotgi) [reply wrangler-core/src/test/java/io/cdap/directives/row/SendToErrorAndContinueTest.java]: removed

File: wrangler-core/src/main/java/io/cdap/directives/row/RecordConditionFilter.java (hunk 17 above)
Comment (nitinmotgi) [reply wrangler-core/src/main/java/io/cdap/directives/row/RecordConditionFilter.java:100]: Was not there, There was no way to convert something to an number, so this path never happened. So, removed it.

**Comment by nitinmotgi** [comment]
@albertshau @CuriousVini PTAL.

File: wrangler-api/src/main/java/io/cdap/wrangler/api/Row.java [hunk 18]
```diff
@@ -170,12 +170,20 @@ public int find(String col) {
   }
 
   /**
-   * @return  Length of the row.
+   * @return  width of the row.
```
Comment (albertshau) [inline wrangler-api/src/main/java/io/cdap/wrangler/api/Row.java:173]: add a `@deprecated` to javadoc and mention what should be used instead.

File: wrangler-core/src/test/java/io/cdap/directives/parser/CsvParserTest.java [hunk 19]
```diff
@@ -67,4 +67,79 @@ public void testHeaders() throws Exception {
...
+      new Row("school_id", "14J456").add("student_id", "13233121").add("last_name", "Diggory")
+        .add("first_name", "Cedric"),
+      new Row("school_id", "14J456").add("student_id", "98786868").add("last_name", "Weasley")
+        .add("first_name", "George"),
+      new Row("school_id", "14J456").add("student_id", "78977876").add("last_name", "Weasley")
+        .add("first_name", "Fred")
+    );
+
+    rows = TestingRig.execute(directives, rows);
+    Assert.assertEquals(6, rows.size());
```
Comment (albertshau) [inline wrangler-core/src/test/java/io/cdap/directives/parser/CsvParserTest.java:107]: can you just do `Assert.assertEquals(expected, rows)`?

File: wrangler-core/src/test/java/io/cdap/directives/parser/CsvParserTest.java [hunk 20]
```diff
@@ -67,4 +67,79 @@ public void testHeaders() throws Exception {
...
+    };
+
+    List<Row> rows = Arrays.asList(
+      new Row("body", "school_id, student_id, last_name, first_name"),
+      new Row("body", "14J456,0033445566,Potter,Jr,Harry"),
+      new Row("body", "14J456,0044333433,Weasley,Ron")
+    );
+
+    rows = TestingRig.execute(directives, rows);
+    Assert.assertEquals(2, rows.size());
```
Comment (albertshau) [inline wrangler-core/src/test/java/io/cdap/directives/parser/CsvParserTest.java:135]: similar comment here

**Review by albertshau — APPROVED** [review]
minor comments, otherwise lgtm. please fix, squash, then merge
